plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/edit.tsx

```diff
@@ -1,25 +1,78 @@
 /**
  * External dependencies
  */
+import clsx from 'clsx';
 import { InspectorControls, useBlockProps } from '@wordpress/block-editor';
 import { PanelBody } from '@wordpress/components';
 import { WC_BLOCKS_IMAGE_URL } from '@woocommerce/block-settings';
 import type { BlockEditProps } from '@wordpress/blocks';
+import { useRef, useState, useEffect } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import { ProductGalleryThumbnailsBlockSettings } from './block-settings';
+import { checkOverflow } from '../../utils';
 import type { ProductGalleryThumbnailsBlockAttributes } from './types';
 
 export const Edit = ( {
 	attributes,
 	setAttributes,
 }: BlockEditProps< ProductGalleryThumbnailsBlockAttributes > ) => {
-	const blockProps = useBlockProps( {
-		className: `wc-block-product-gallery-thumbnails wc-block-product-gallery-thumbnails--number-of-thumbnails-${ attributes.numberOfThumbnails }`,
+	const { thumbnailSize } = attributes;
+	const minSize = 10;
+	const maxSize = 50;
+	const defSize = 33;
+
+	const scrollableRef = useRef< HTMLDivElement >( null );
+	const [ overflowState, setOverflowState ] = useState( {
+		overflowBottom: false,
+		overflowRight: false,
 	} );
 
+	useEffect( () => {
+		const scrollableElement = scrollableRef.current;
+		if ( ! scrollableElement ) return;
+
+		// Create a ResizeObserver to watch for layout changes
+		const resizeObserver = new ResizeObserver( () => {
+			const overflow = checkOverflow( scrollableElement );
+			setOverflowState( overflow );
+		} );
+
+		// Observe both the scrollable element and its parent for size changes
+		resizeObserver.observe( scrollableElement );
+		if ( scrollableElement.parentElement ) {
+			resizeObserver.observe( scrollableElement.parentElement );
+		}
+
+		// Initial check
+		const overflow = checkOverflow( scrollableElement );
+		setOverflowState( overflow );
+
+		return () => {
+			resizeObserver.disconnect();
+		};
+	}, [ thumbnailSize ] ); // Re-run when thumbnailSize changes as it affects layout
+
+	const thumbnailSizeValue =
+		Math.min(
+			Math.max( Number( thumbnailSize.replace( '%', '' ) ), minSize ),
+			maxSize
+		) || defSize;
```

Comment:
Isn't `minSize > thumbnailSize > maxSize` guaranteed from the block control settings?

Reply:
It is but only on blur. During the input any value is passed which causes a huge layout jumps. For example when you want to input `15`, when you remove everything, the value was a default (33), you input `1` and it goes to `1`, then `15`. Or if you accidentally put higher number then it's super big. I thought having some validation would be helpful.

I moved that logic to `onChange` in `plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/block-settings/index.tsx` though.

Reply:
Maybe switching back to `RangeControl` makes more sense and would avoid this complexity?

Reply:
I'd stick with `UnitControl` to keep it consistent across Editor. This is the control used to input sizes.